### PR TITLE
Clip update for onnx

### PR DESCRIPTION
**Pull request by kahmed10**
#439

### CODE REVIEW

**Comment by pfultz2** [comment]
Why dont we change the clip operator to take the min and max as input values?

**Comment by kahmed10** [comment]
> Why dont we change the clip operator to take the min and max as input values?

Made changes to accept min and max as input args

File: src/onnx/onnx.cpp [hunk 1]
```diff
@@ -325,16 +325,32 @@ struct onnx_parser
                                const attribute_map& attributes,
                                std::vector<instruction_ref> args)
     {
-        op::clip op;
+        auto input_lens = args[0]->get_shape().lens();
+        instruction_ref min_arg;
```
Comment (scxiao) [inline src/onnx/onnx.cpp]: If the operator has only one input and there is not "min" or "max", the min_arg and max_arg will be empty here. The logic would be something like 
        if(args.size() > 2)
        {
            max_arg = args[2];
        }
        else
        {
            float max_val = std::numeric_limits<float>::max();
            if(contains(attributes, "max"))
            {
                max_val = parse_value(attributes.at("max")).at<float>();
            }
            max_arg = prog.add_literal(max_val);
        }

        if(args.size() > 1)
        {
            min_arg = args[1];
        }
        else
        {
            float min_val = std::numeric_limits<float>::min();
            if(contains(attributes, "min"))
            {
                min_val = parse_value(attributes.at("min")).at<float>();
            }
            min_arg = prog.add_literal(min_val);
        }

If there is no min or max value provided, the default value is used. (Actually, for this case, the clip operator can be removed.)

File: src/onnx/onnx.cpp [hunk 2]
```diff
@@ -325,16 +325,32 @@ struct onnx_parser
...
+            max_arg       = prog.add_literal(max_val);
         }
         if(contains(attributes, "min"))
         {
-            op.min_val = parse_value(attributes.at("min")).at<float>();
+            float min_val = parse_value(attributes.at("min")).at<float>();
+            min_arg       = prog.add_literal(min_val);
         }
-        return prog.add_instruction(op, std::move(args));
+        min_arg = prog.add_instruction(op::multibroadcast{input_lens}, min_arg);
```
Comment (scxiao) [inline src/onnx/onnx.cpp]: I am wondering why we need to multibroadcast? We can just use the index 0 to get the min/max value, right?

Comment (kahmed10) [reply src/onnx/onnx.cpp]: I added a multibroadcast to reuse nary() in the device function without adding more complexity.

File: src/onnx/onnx.cpp (hunk 1 above)
Comment (kahmed10) [reply src/onnx/onnx.cpp]: I agree that I need to fix the logic. I am also debating on whether to assign default values to min/max or simply using a max operator when no max provided, and an identity if neither operators are provided.

Comment (scxiao) [reply src/onnx/onnx.cpp]: Yes, I am fine with not adding the operator if neither of min/max is used.

Comment (kahmed10) [reply src/onnx/onnx.cpp]: fixed

Comment (kahmed10) [reply src/onnx/onnx.cpp]: fixed

File: src/onnx/onnx.cpp [hunk 3]
```diff
@@ -330,16 +330,48 @@ struct onnx_parser
...
+        if(min_used)
+            min_arg = prog.add_instruction(op::multibroadcast{input_lens}, min_arg);
+
+        if(max_used)
+            max_arg = prog.add_instruction(op::multibroadcast{input_lens}, max_arg);
+
+        if(min_used and max_used)
+            return prog.add_instruction(op::clip{}, args[0], min_arg, max_arg);
+        if(min_used)
+            return prog.add_instruction(op::max{}, args[0], min_arg);
```
Comment (pfultz2) [inline src/onnx/onnx.cpp]: Why is the max op being used with the min_arg?

Comment (kahmed10) [reply src/onnx/onnx.cpp]: If only the minimum value is clipped, you can simply call max(input, min_arg), which will prevent the input from going less than the minimum value.

Comment (pfultz2) [reply src/onnx/onnx.cpp]: I see its because we are clipping the min value.

**Pull request merged**